sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -135,154 +157,212 @@ namespace Azure { namespace Core { namespace Http {
+    static std::string EncodeFragment(const std::string& fragment);
+    static std::string EncodeImpl(
+        const std::string& source,
+        const std::function<bool(int)>& should_encode);
+
+    void StartRetry()
+    {
+      m_retryModeEnabled = true;
+      m_retryQueryParameters.clear();
+    }
 
-    /**
-     * Will check if there are any query parameter in url looking for symbol '?'
-     * If it is found, it will insert query parameters to m_queryParameters internal field
-     * and remove it from url
-     */
-    const std::string SaveAndRemoveQueryParameter(std::string const& url)
+    void StopRetry()
     {
+      m_retryModeEnabled = false;
+      m_retryQueryParameters.clear();
+    }
 
-      const auto firstPosition = std::find(url.begin(), url.end(), '?');
-      if (firstPosition == url.end())
-      {
-        return url; // not query parameters
-      }
+    std::string GetEncodedQuery() const;
 
-      auto position = firstPosition; // position of symbol ?
-      while (position != url.end())
-      {
-        ++position; // skip over the ? or &
-        const auto nextPosition = std::find(position, url.end(), '&');
-        const auto equalChar = std::find(position, nextPosition, '=');
-        auto valueStart = equalChar;
-        if (valueStart != nextPosition)
-        {
-          ++valueStart; // skip = symbol
-        }
-
-        // Note: if there is another = symbol before nextPosition, it will be part of the
-        // paramenter value. And if there is not a ? symbol, we add empty string as value
-        this->m_queryParameters.insert(std::pair<std::string, std::string>(
-            std::string(position, equalChar), std::string(valueStart, nextPosition)));
-
-        position = nextPosition;
-      }
+  public:
+    // Create empty Url instance. Usually for building Url from scratch
+    Url() {}
+
+    // Create Url from an url-encoded string. Usually from pre-built url with query parameters (like
+    // SaS) url is expected to be already url-encoded.
+    explicit Url(const std::string& url);
+
+    /************* Builder Url functions ****************/
+    /******** API for building Url from scratch. Override state ********/
+
+    void SetScheme(const std::string& scheme) { m_scheme = scheme; }
 
-      return std::string(url.begin(), firstPosition);
+    void SetHost(const std::string& host, bool isHostEncoded = false)
+    {
+      m_host = isHostEncoded ? host : EncodeHost(host);
     }
 
-  public:
-    URL(std::string const& url);
-    void AppendPath(std::string const& path)
+    void SetPort(uint16_t port) { m_port = port; }
+
+    void SetPath(const std::string& path, bool isPathEncoded = false)
+    {
+      m_encodedPath = isPathEncoded ? path : EncodePath(path);
+    }
+
+    void SetFragment(const std::string& fragment, bool isFragmentEncoded = false)
+    {
+      m_fragment = isFragmentEncoded ? fragment : EncodeFragment(fragment);
```

Comment:
Yes. Storage uses this uri builder to construct sas token, where scheme, path, host and fragment are all empty.

Reply:
Interesting. Why would you call `SetFragment` in that case?

Reply:
`SetFragment` is not used at all. It exists just for completeness. People may find it weird if we can set host, we can set path, we can set query parameters, but we cannot set fragment.

Reply:
But we can remove it if you don't like it. I'm totally fine.

Reply:
I get your point, but in this case, I think adding it once we need it is a safer stance than having to remove or break it later, especially since I am not sure how many azure services use the fragment piece. I don't see the SDK's goal to create a complete, general-purpose Url class/APIs, but rather what is needed to support on-boarding the various service clients.

As long as we are confident in the design, and we make sure it allows such extensions in the future.